bplustree_test.go

```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
+	for i := 0; i < len(result); i++ {
+		//fmt.Println(reflect.TypeOf(result[i]))
+		actualVal := result[i].(*TestingElemStr).val
+		if actualVal != expectedVal {
+			t.Errorf("At index %d: expected: %d, got %v", i, expectedVal, actualVal)
+			t.FailNow()
+		}
+		// and increment by 1.
+		expectedVal++
+	}
+}
+
+func TestSearchStr6(t *testing.T) {
+
+	// case 6. use of evaluator with maxElems = 10, such that
+	// evaluator yields more than 10 elements. We should still
+	// get 10 elements.
+
+	testingKey := makeTestingKeyStr(100)
+	ss := SearchSpecifier{searchKey: testingKey, direction: Both,
+		maxElems: 10, evaluator: func(k1 Key, k2 Key) bool {
+			result := k1.Compare(k2)
+			if result < 10 && result > -10 {
+				return true
+			}
+
+			return false
+		},
+	}
+
+	result, err := newtree.Search(ss)
+	if err != nil {
+		t.Errorf("Expected to find the key: %v", ss.searchKey)
+		t.FailNow()
+	}
+	if result == nil {
+		t.Errorf("Expected to get result but nil: %v", result)
+		t.FailNow()
+	}
+
+	if len(result) < ss.maxElems || len(result) > ss.maxElems+1 {
+		t.Errorf("Expected to find atleast %d results, got: %d", ss.maxElems, len(result))
+		t.FailNow()
+	}
+
+	// Note that our values are +10 of the key.
+	startKey := makeTestingKeyStr(strHash(testingKey) - 5)
+	expectedVal := int(strHash(startKey) + 10)
+	for i := 0; i < len(result); i++ {
+		//fmt.Println(reflect.TypeOf(result[i]))
+		actualVal := result[i].(*TestingElemStr).val
+		if actualVal != expectedVal {
+			t.Errorf("At index %d: expected: %d, got %v", i, expectedVal, actualVal)
+			t.FailNow()
+		}
+		// and increment by 1.
+		expectedVal++
+	}
+}
+
+func TestSearchStr7(t *testing.T) {
+
+	// case 7. use of evaluator with maxElems = 10, such that
+	// evaluator yields less than 10 elements. We should only
+	// get less than 10 elements.
+
+	testingKey := makeTestingKeyStr(100)
+	ss := SearchSpecifier{searchKey: testingKey, direction: Both,
+		maxElems: 10, evaluator: func(k1 Key, k2 Key) bool {
+			result := k1.Compare(k2)
+			if result < 3 && result > -3 {
+				return true
+			}
+
+			return false
+		},
+	}
+
+	result, err := newtree.Search(ss)
+	if err != nil {
+		t.Errorf("Expected to find the key: %v", ss.searchKey)
+		t.FailNow()
+	}
+	if result == nil {
+		t.Errorf("Expected to get result but nil: %v", result)
+		t.FailNow()
+	}
+
+	if len(result) != 5 {
+		t.Errorf("Expected to find 5 results, got: %d", len(result))
+		t.FailNow()
+	}
+
+	// Note that our values are +10 of the key.
+	startKey := makeTestingKeyStr(strHash(testingKey) - 2)
+	expectedVal := int(strHash(startKey) + 10)
+	for i := 0; i < len(result); i++ {
+		//fmt.Println(reflect.TypeOf(result[i]))
+		actualVal := result[i].(*TestingElemStr).val
+		if actualVal != expectedVal {
+			t.Errorf("At index %d: expected: %d, got %v", i, expectedVal, actualVal)
+			t.FailNow()
+		}
+		// and increment by 1.
+		expectedVal++
+	}
+}
+
+func TestRemoveStr1(t *testing.T) {
+	fmt.Println("Starting remove in order tests")
```

Comment:
are you expecting the test functions to run in order?  may not be a safe assumption if someone adds code in the file or moves things around.

Reply:
Unless I write a giant function which does everything.. it has to be this way.. can't expect to remove things if they aren't 'insert'ed. And if tests are added in order that it breaks something then the test will break right away and whoever adds new test would know. And I just don't want to have 1 single function with all functionality. During multiple remove tests through.. I am reinserting the keys,, so no ordering is needed.. I could make each testcase be self sufficient (unlike only some of them which are), but then inserting 100K entries for each test case becomes an overkill..